### PR TITLE
Rust 2018 migration

### PR DESCRIPTION
**Pull request by mati865**
Clippy being nightly-only crate can easily migrate to new Rust edition and start using new features.

This PR gives insight of required changes and is was created to open discussion about the migration.
I think it looks mostly fine although I'm not sure about setting edition in `Cargo.toml` just yet.

### CODE REVIEW

**Comment by mati865** [comment]
Meh
```
$ npm install remark-cli remark-lint
npm ERR! code E429
npm ERR! 429 Too Many Requests: remark-lint@latest
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2018-05-30T09_22_58_988Z-debug.log
```

**Comment by oli-obk** [comment]
Yea, we're ddosing npm -.- there's a commit in some PR, but it hasn't been merged yet

File: clippy_lints/src/new_without_default.rs [hunk 1]
```diff
@@ -153,7 +153,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NewWithoutDefault {
 }
 
 fn create_new_without_default_suggest_msg(ty: Ty) -> String {
-    #[rustfmt_skip]
+    #[cfg_attr(rustfmt, rustfmt_skip)]
```
Comment (oli-obk) [inline clippy_lints/src/new_without_default.rs]: I think that should be `#[rustfmt::skip]` without the `cfg_attr`

**Comment by oli-obk** [comment]
This is awesome! @killercup and I were talking about this at some point this in the last 24h. And now there's a PR :heart:

File: clippy_lints/src/new_without_default.rs (hunk 1 above)
Comment (mati865) [reply clippy_lints/src/new_without_default.rs]: I also thought about it but decided to keep more consistent with current code base.
<details>
<summary>click here for `rg` output</summary>

```
$ rg "rustfmt" .     
pre_publish.sh
23:echo "remember to add a git tag and running 'cargo test' before committing the rustfmt changes"

tests/trim_multiline.rs
16:#[cfg_attr(rustfmt, rustfmt_skip)]
41:#[cfg_attr(rustfmt, rustfmt_skip)]

tests/needless_continue_helpers.rs
8:#[cfg_attr(rustfmt, rustfmt_skip)]
26:#[cfg_attr(rustfmt, rustfmt_skip)]
38:#[cfg_attr(rustfmt, rustfmt_skip)]
57:#[cfg_attr(rustfmt, rustfmt_skip)]
73:#[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/new_without_default.rs
156:    #[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/loops.rs
1781:#[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/non_expressive_names.rs
91:#[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/methods.rs
1824:#[cfg_attr(rustfmt, rustfmt_skip)]
1834:#[cfg_attr(rustfmt, rustfmt_skip)]
1868:#[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/matches.rs
200:#[cfg_attr(rustfmt, rustfmt_skip)]

clippy_lints/src/lib.rs
211:#[cfg_attr(rustfmt, rustfmt_skip)]
```
</details>

Comment (oli-obk) [reply clippy_lints/src/new_without_default.rs]: oh ^^

**Comment by oli-obk** [comment]
The future is now. Let's see how this goes

**Comment by mati865** [comment]
@nikomatsakis might be interested to see it.

**Pull request merged**